integration/integration_test.go

```diff
@@ -4937,6 +4938,138 @@ func testExecEvents(t *testing.T, suite *integrationTestSuite) {
 	}
 }
 
+func testSessionStartContainsAccessRequest(t *testing.T, suite *integrationTestSuite) {
+	accessRequestsKey := "access_requests"
+	requestedRoleName := "requested-role"
+	userRoleName := "user-role"
+
+	tr := utils.NewTracer(utils.ThisFunction()).Start()
+	defer tr.Stop()
+
+	lsPath, err := exec.LookPath("ls")
+	require.NoError(t, err)
+
+	// Creates new teleport cluster
+	main := suite.newTeleport(t, nil, true)
+	defer main.StopAll()
+
+	ctx := context.Background()
+	// Get auth server
+	authServer := main.Process.GetAuthServer()
+
+	// Create new request role
+	requestedRole, err := services.NewRole(requestedRoleName, services.RoleSpecV3{
+		Options: services.RoleOptions{},
+		Allow:   services.RoleConditions{},
+	})
+	require.NoError(t, err)
+
+	err = authServer.UpsertRole(ctx, requestedRole)
+	require.NoError(t, err)
+
+	// Create user role with ability to request role
+	userRole, err := services.NewRole(userRoleName, services.RoleSpecV3{
+		Options: services.RoleOptions{},
+		Allow: services.RoleConditions{
+			Request: &services.AccessRequestConditions{
+				Roles: []string{requestedRoleName},
+			},
+		},
+	})
+	require.NoError(t, err)
+
+	err = authServer.UpsertRole(ctx, userRole)
+	require.NoError(t, err)
+
+	user, err := services.NewUser(suite.me.Username)
+	user.AddRole(userRole.GetName())
+	require.NoError(t, err)
+
+	watcher, err := authServer.NewWatcher(ctx, types.Watch{
+		Kinds: []types.WatchKind{
+			{Kind: types.KindUser},
+		},
+	})
+	require.NoError(t, err)
+	defer watcher.Close()
+
+	select {
+	case <-time.After(time.Second * 30):
+		t.Fatalf("Timeout waiting for event.")
+	case event := <-watcher.Events():
+		if event.Type != types.OpInit {
+			t.Fatalf("Unexpected event type.")
+		}
+		require.Equal(t, event.Type, types.OpInit)
+	case <-watcher.Done():
+		t.Fatal(watcher.Error())
+	}
+
+	// Update user
+	err = authServer.UpsertUser(user)
+	require.NoError(t, err)
+
+	WaitForResource(t, watcher, user.GetKind(), user.GetName())
+
+	req, err := services.NewAccessRequest(suite.me.Username, requestedRole.GetMetadata().Name)
+	require.NoError(t, err)
+
+	accessRequestID := req.GetName()
+
+	err = authServer.CreateAccessRequest(context.TODO(), req)
+	require.NoError(t, err)
+
+	err = authServer.SetAccessRequestState(context.TODO(), types.AccessRequestUpdate{
+		RequestID: accessRequestID,
+		State:     types.RequestState_APPROVED,
+	})
+	require.NoError(t, err)
```

Comment:
Nit: IIRC the reissue operation reads from cache, so it might be worth doing a `WaitForResource` after the creation and update of the access request as well.

Reply:
Added!